hs-abci-examples/nameservice/test/Nameservice/Test/E2ESpec.hs

```diff
@@ -0,0 +1,244 @@
+                                                       Name (..), SetName (..),
+                                                       Whois (..))
+import           Nameservice.Modules.Token            (Amount (..))
+import qualified Network.ABCI.Types.Messages.Response as Response
+import qualified Network.Tendermint.Client            as RPC
+import           Proto3.Suite                         (Message,
+                                                       toLazyByteString)
+import           Servant.API                          ((:<|>) (..))
+import           Tendermint.SDK.Codec                 (HasCodec (..))
+import           Tendermint.SDK.Crypto                (Secp256k1,
+                                                       addressFromPubKey)
+import           Tendermint.SDK.Query.Client          (ClientResponse (..),
+                                                       genClient)
+import           Tendermint.SDK.Query.Types           (QueryArgs (..))
+import           Tendermint.SDK.Types.Address         (Address (..))
+import           Tendermint.SDK.Types.Transaction     (RawTransaction (..),
+                                                       signRawTransaction)
+import           Test.Hspec
+
+spec :: Spec
+spec = do
+  let satoshi = Name "satoshi"
+      addr1 = userAddress user1
+      privateKey1 = userPrivKey user1
+      addr2 = userAddress user2
+      privateKey2 = userPrivKey user2
+
+  beforeAll (do faucetAccount user1; faucetAccount user2) $
+    describe "Nameservice Spec" $ do
+      it "Can query /health to make sure the node is alive" $ do
+        resp <- runRPC RPC.health
+        resp `shouldBe` RPC.ResultHealth
+
+      it "Can query account balances" $ do
+        let queryReq = defaultReqWithData addr1
+        ClientResponse{clientResponseData = foundAmount} <- runRPC $ getBalance queryReq
+        foundAmount `shouldBe` Amount 1000
+
+      it "Can create a name" $ do
+        let msg = BuyName 0 satoshi "hello world" addr1
+            rawTx = mkSignedRawTransactionWithRoute "nameservice" privateKey1 msg
+        ensureTxCommitResponseCode rawTx 0
+
+      it "Can query for a name" $ do
+        let queryReq = defaultReqWithData satoshi
+        ClientResponse{clientResponseData = foundWhois} <- runRPC $ getWhois queryReq
+        whoisValue foundWhois `shouldBe` "hello world"
+        whoisOwner foundWhois `shouldBe` addr1
+        whoisPrice foundWhois `shouldBe` 0
+
+      it "Can query for a name that doesn't exist" $ do
+        let nope = Name "nope"
+            queryReq = defaultReqWithData nope
+        ClientResponse{ clientResponseData = emptyWhois
+                      , clientResponseRaw
+                      } <- runRPC $ getWhois queryReq
+        let queryRespCode = clientResponseRaw ^. Response._queryCode
+        -- storage failure
+        queryRespCode `shouldBe` 1
+        -- empty whois (defaults)
+        whoisPrice emptyWhois `shouldBe` 0
```

Comment:
Am I understanding that the protobuf messages can be decoded from an empty bytestring, similar to how the EVM returns empty structs even if they aren't found in a mapping?

Reply:
Yeah, I think the decoding library behaves that way. Each field of a decoded message is assumed to have a `HasDefault` instance.